firebase-app-distribution/src/main/java/com/google/firebase/appdistribution/CheckForUpdateClient.java

```diff
@@ -165,4 +193,16 @@ private long getInstalledAppVersionCode(Context context) throws FirebaseAppDistr
     }
     return PackageInfoCompat.getLongVersionCode(pInfo);
   }
+
+  @VisibleForTesting
+  String extractApkCodeHash(PackageInfo packageInfo) {
+    File sourceFile = new File(packageInfo.applicationInfo.sourceDir);
+
+    String key =
+        String.format(
+            Locale.ENGLISH, "%s.%d", sourceFile.getAbsolutePath(), sourceFile.lastModified());
+
+    cachedCodeHashes.putIfAbsent(key, calculateApkInternalCodeHash(sourceFile));
```

Comment:
Does this `calculateApkInternalCodeHash(sourceFile)` run every time? I re-read the documentation for putIfAbsent, and I am concerned it might call the calculateApkInternalCodeHash even if the key exists in the map.

Could we add a test to make sure we don't call calculateApkInternalCodeHash if the file exists in the cached hash map?

Reply:
After writing a unit test it does recalculate it apparently, I am going to rewrite the logic so that it does it only once